generator/internal/parser/httprule/http_rule_parser_test.go

```diff
@@ -0,0 +1,180 @@
+		{"/v1/*/foo", template(literal("v1"), match(), literal("foo")), ""},
+		{"/v1/**/foo", template(literal("v1"), matchR(), literal("foo")), ""},
+		{"/foo:bar", templateV(segments(literal("foo")), verb("bar")), ""},
+		{"/foo/{bar}", template(literal("foo"), varr("bar")), ""},
+		{"/foo/{bar.baz}", template(literal("foo"), varr("bar", "baz")), ""},
+		{"/foo/{bar=baz}", template(literal("foo"), varrs(ids("bar"), segments(literal("baz")))), ""},
+		{"/foo/{bar=*}", template(literal("foo"), varrs(ids("bar"), segments(match()))), ""},
+		{"/foo/{bar=*}/baz", template(literal("foo"), varrs(ids("bar"), segments(match())), literal("baz")), ""},
+		{"/foo/{bar=*}/baz:qux", templateV(segments(literal("foo"), varrs(ids("bar"), segments(match())), literal("baz")), verb("qux")), ""},
+		{"foo", nil, "path must start with slash"},
+		{"/", nil, "path cannot end with slash"},
+		{"/foo/", nil, "path cannot end with slash"},
+
+		// the following test is failing because * is a sub-delimiter, which is allowed in the LITERAL segment
+		//{"/foo/***/bar", nil, "wildcard literal cannot exceed two *"},
+
+		{"/foo/:bar", nil, "verb cannot come after slash"},
+		{"/foo:bar/baz", nil, "verb must be the last segment"},
+		{":foo", nil, "verb cannot be the first segment"},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.path, func(t *testing.T) {
+			got, err := Parse(tc.path)
+			if tc.want != nil {
+				if err != nil {
+					t.Fatalf("expected no error, got: %v", err)
+				}
+				if got == nil {
+					t.Fatalf("expected path template for %s, got nil", tc.path)
+				}
+				if diff := cmp.Diff(tc.want, got, cmpopts.EquateEmpty()); diff != "" {
+					t.Fatalf("failed parsing path [%s] (-want, +got):\n%s", tc.path, diff)
+				}
+			} else {
+				if err == nil {
+					t.Fatalf("Parse(%s) succeeded, want error: %s", tc.path, tc.explanation)
+				}
+			}
+		})
+	}
+}
+
+func Test_parseSegments(t *testing.T) {
+	tests := []struct {
+		path        string
+		want        []*Segment
+		explanation string
+	}{
+		{"foo", []*Segment{{Literal: newLiteral("foo")}}, ""},
+		{"foo/bar", []*Segment{{Literal: newLiteral("foo")}, {Literal: newLiteral("bar")}}, ""},
+		{"v1/*/foo", []*Segment{{Literal: newLiteral("v1")}, {Match: &Match{}}, {Literal: newLiteral("foo")}}, ""},
+		{"v1/**/foo", []*Segment{{Literal: newLiteral("v1")}, {MatchRecursive: &MatchRecursive{}}, {Literal: newLiteral("foo")}}, ""},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.path, func(t *testing.T) {
+			if tc.want != nil {
+				got, err := parseSegments(tc.path)
+				if err != nil {
+					t.Fatalf("expected no error, got: %v", err)
+				}
+				if diff := cmp.Diff(tc.want, got, cmpopts.EquateEmpty()); diff != "" {
+					t.Fatalf("failed parsing path [%s] (-want, +got):\n%s", tc.path, diff)
+				}
+			} else {
+				_, err := parseSegments(tc.path)
+				if err == nil {
+					t.Fatalf("Parse(%s) succeeded, want error: %s", tc.path, tc.explanation)
+				}
+			}
+
+		})
+	}
+
+}
+
```

Comment:
constructing the `want` data structure is quite verbose, but I am not sure using these helper methods makes it easier to read